### DISCUSSION TITLE
'resolveMagicCall()' classify wrong? snake_case fields cannot be found

### DISCUSSION THREAD

**Discussion by nsd4npo** [doctrine/doctrine2#10938] [Support Questions]
Greetings,

I'm using a custom function within my `defaultRepository` extender class called `findOneOrCreateByField($fieldName, $fieldValue)`. Here's a snippet of how it works:
```
$inflector = InflectorFactory::create()->build();
$field = $inflector->classify($field);
$finder = "findOneBy" . $field;
$entity = $this->$finder($fieldValue);
```

However, I encountered an issue when using it with the parameters `testTime` and `\DateTime("now")`, which resulted in a "field not found" error. After some debugging, I discovered the source of the problem in this (shortened) code block:

```
private function resolveMagicCall(string $method, string $by, array $arguments)
    $fieldName = lcfirst(self::$inflector->classify($by));

    if (!($this->_class->hasField($fieldName) || $this->_class->hasAssociation($fieldName))) {
        throw InvalidMagicMethodCall::becauseFieldNotFoundIn(
```
https://github.com/doctrine/orm/blob/2.16.x/lib/Doctrine/ORM/EntityRepository.php#L358-L360

In this context, the code uses classify on the field name, resulting in `TestTime`. The subsequent `hasField` check fails because it's looking for a `test_time` field instead of `TestTime`.

![grafik](https://github.com/doctrine/orm/assets/143995254/b5dbd732-26a3-41a2-afd1-391fda907eda)

My conclusion is that for the `hasField` check, it should use snake_case or tableize, while for the `hasAssociation` check, using classify is appropriate.

Workaround in findOneOrCreateByField:
```
$field = $inflector->tableize($field);
$entity = $this->findOneBy([$field => $fieldValue]);
```

**Comment by derrabus** [comment]
tbh, your "workaround" looks better than the actual code. Assembling a method name just to hit the correct magic call resolution does not make sense to me.

**Reply by reloxx13** [reply]
There are many ways to lead to Rome 😉, but the magic resolver is still buggy as soon as you have fields with underscores.   
A direct call of `$this->findOneByTestTime("2023-09-11")` wouldn't work either.

OT: I extended another function that used the magic resolver method for only one specific field. I didn't want to change the original code more than necessary. However, the 'workaround' works and will remain in place as an improvement.
